src/passes/OptimizeInstructions.cpp

```diff
@@ -1270,12 +1291,30 @@ struct OptimizeInstructions
     return binary;
   }
 
-  template<typename T> Expression* optimizePowerOf2URem(Binary* binary, T c) {
-    static_assert(std::is_same<T, uint32_t>::value ||
-                    std::is_same<T, uint64_t>::value,
+  template<typename T> Expression* optimizePowerOf2FDiv(Binary* binary, T c) {
+    //
+    // x / C_pot    =>   x * (C_pot ^ -1)
+    //
+    // Explanation:
+    // Floating point numbers are represented as:
+    //    ((-1) ^ sign) * (2 ^ (exp - bias)) * (1 + significand)
+    //
+    // If we have power of two numbers, then the mantissa (significand)
+    // is all zeros. Let's focus on the exponent, ignoring the sign part:
+    //    (2 ^ (exp - bias))
+    //
+    // and for inverted power of two floating point:
+    //     1.0 / (2 ^ (exp - bias))   ->   2 ^ -(exp - bias)
```

Comment:
Hmm, interesting...

I think the number still must be written as `2 ^ (something - bias)` for the same fixed `bias` as always. So we need to do
```
 1.0 / (2 ^ (exp - bias))   ->   2 ^ -(exp - bias)   ==   2 ^ (-exp + 2*bias - bias)
```
That is, the new exponent is `-exp + 2*bias`. Is that always a valid exponent?

In a double, the exponent is 11 bits and the exponent of a non-NaN, non-inf number is in the range `[1, 2046]` (since `0`, `2047` are reserved). That means that we need the new exponent to fit in that range,
```
-exp + 2*bias = -exp + 2046

and we need

1 <= -exp + 2046 <= 2046
-2046 <= exp - 2046 <= -1
0 <= exp <= 2045
```

If my math is right, there is a possible problem with an original exponent of 2046, the maximum (non-NaN, non-inf) one? (and the same may happen at the edge for floats, if so?)

Side note, we should improve the fuzzer to emit random floats in relevant patterns for things like this. That is, with some probability to pick a power of 2 number.

Reply:
Also simple check in godot show that's `x / 0x0.8p-126` != `x * 0x1p127`:
https://godbolt.org/z/YMWrfM

Reply:
As I expected, the problem is most likely in the border between normal and denormalized numbers for f32

Reply:
@MaxGraey Not sure what to look at on the godbolt site - I see the output wasm, but how does it show if the optimization is invalid or not?

Do you agree with what I said above, that this optimization is invalid on that corner case?

Reply:
Gobolt's example related to changes which exposed in #3237 where doubles and floats handled the same routines. It seems that proposed refactoring isn't working as we expected. Current implementation presented in this PR (where float and double handled differently) works more correct for 32-bit floats

Reply:
if floats promote to doubles, this case `x / 0x0.8p-126f` rounded to `x / double(0x1p-127)` and pass `isPower2Float` guard. So I just propose stay with current more complex but more precise (valid) version. And probably I'll close #3237

Reply:
@MaxGraey what about my mathematical computation at the beginning of this comment chain, I don't think you've responded to it?

My comment is not about #3237 - it is about this PR, what is on master now.

Reply:
Regarding computations. For doubles we have exponent have [−1022 .. +1023] range but in this PR I limit it to [-1022..+1022] inside `isPower2Float` routine (See MIN_POT and MAX_POT values). So negative of exponent should just reverse this range to [+1022..-1022]. I also made sure that edge cases are really valid in practice: 
`0x1p-1022 == 1.0 / 0x1p1022`
`1.0 / 0x1p-1022 == 0x1p1022`
https://godbolt.org/z/Kafqsh

Reply:
I see, thanks, you limited the range of the exponents in the other file to avoid this... sounds good. But it is a little surprising, since the optimization here does not mention you depend on that. So the description here is confusing I think, let's update it please.

Reply:
Sure, could I update that comment in some already existing PR? Don't see reason create separate pr for such minor change